File: lib/core/catalog/interfaces/catalog_service.dart
import 'package:sertton/core/catalog/dtos/brand_dto.dart';
import 'package:sertton/core/catalog/dtos/category_dto.dart';
import 'package:sertton/core/catalog/dtos/collection_dto.dart';
import 'package:sertton/core/catalog/dtos/product_dto.dart';
import 'package:sertton/core/global/responses/pagination_response.dart';
import 'package:sertton/core/global/responses/rest_response.dart';

abstract class CatalogService {
  Future<RestResponse<PaginationResponse<ProductDto>>> fetchProducts({
    int page = 1,
    String? categoryId,
    List<String> brandsIds = const [],
    String? query,
  });
  Future<RestResponse<ProductDto>> fetchProduct(String productId);
  Future<RestResponse<List<CollectionDto>>> fetchCollections();
  Future<RestResponse<List<CategoryDto>>> fetchCategories();
  Future<RestResponse<List<BrandDto>>> fetchBrands();
  Future<RestResponse<List<ProductDto>>> fetchProductsByCollection(
    String collectionId,
  );
  Future<RestResponse<List<ProductDto>>> fetchSimiliarProducts(
    String productId,
  );
}

File: sertton-react-native-app/src/core/catalog/interfaces/catalog-service.ts
import type { Brand, Category, Collection, Product } from "@/core/catalog/entities"
import type { Pagination } from "@/core/shared/responses/pagination"
import type { RestResponse } from "@/core/shared/responses/rest-response"

export interface FetchProductsParams {
  readonly brandsIds?: readonly string[]
  readonly categoryId?: string
  readonly page?: number
  readonly query?: string
}

export interface CatalogService {
  fetchBrands(): Promise<RestResponse<readonly Brand[]>>
  fetchCategories(): Promise<RestResponse<readonly Category[]>>
  fetchCollections(): Promise<RestResponse<readonly Collection[]>>
  fetchProduct(productId: string): Promise<RestResponse<Product>>
  fetchProducts(params?: FetchProductsParams): Promise<RestResponse<Pagination<Product>>>
  fetchProductsByCollection(collectionId: string): Promise<RestResponse<readonly Product[]>>
  fetchSimilarProducts(productId: string): Promise<RestResponse<readonly Product[]>>
}

export type ICatalogService = CatalogService
